src/promql/src/extension_plan/union_distinct_on.rs

```diff
@@ -0,0 +1,576 @@
+    right: HashedDataFut,
+    /// Include time index
+    compare_keys: Vec<usize>,
+    output_schema: SchemaRef,
+    metric: BaselineMetrics,
+}
+
+impl UnionDistinctOnStream {
+    fn poll_impl(&mut self, cx: &mut Context<'_>) -> Poll<Option<<Self as Stream>::Item>> {
+        // resolve the right stream
+        let right = match self.right {
+            HashedDataFut::Pending(ref mut fut) => {
+                let right = ready!(fut.as_mut().poll(cx))?;
+                self.right = HashedDataFut::Ready(right);
+                let HashedDataFut::Ready(right_ref) = &mut self.right else {
+                    unreachable!()
+                };
+                right_ref
+            }
+            HashedDataFut::Ready(ref mut right) => right,
+            HashedDataFut::Empty => return Poll::Ready(None),
+        };
+
+        // poll left and probe with right
+        let next_left = ready!(self.left.poll_next_unpin(cx));
+        match next_left {
+            Some(Ok(left)) => {
+                // observe left batch and return it
+                right.update_map(&left)?;
+                Poll::Ready(Some(Ok(left)))
+            }
+            Some(Err(e)) => Poll::Ready(Some(Err(e))),
+            None => {
+                // left stream is exhausted, so we can send the right part
+                let right = std::mem::replace(&mut self.right, HashedDataFut::Empty);
+                let HashedDataFut::Ready(data) = right else {
+                    unreachable!()
+                };
+                Poll::Ready(Some(data.finish()))
+            }
+        }
+    }
+}
+
+impl RecordBatchStream for UnionDistinctOnStream {
+    fn schema(&self) -> SchemaRef {
+        self.output_schema.clone()
+    }
+}
+
+impl Stream for UnionDistinctOnStream {
+    type Item = DataFusionResult<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.poll_impl(cx)
+    }
+}
+
+/// Simple future state for [HashedData]
+enum HashedDataFut {
+    /// The result is not ready
+    Pending(BoxFuture<'static, DataFusionResult<HashedData>>),
+    /// The result is ready
+    Ready(HashedData),
+    /// The result is taken
+    Empty,
+}
+
+/// ALL input batches and its hash table
+struct HashedData {
+    // TODO(ruihang): use `JoinHashMap` instead after upgrading to DF 34.0
+    /// Hash table for all input batches. The key is hash value, and the value
+    /// is the index of `bathc`.
+    hash_map: HashMap<u64, usize>,
+    /// Output batch.
+    batch: RecordBatch,
+    /// The indices of the columns to be hashed.
+    hash_key_indices: Vec<usize>,
+    random_state: RandomState,
+}
+
+impl HashedData {
+    pub async fn new(
+        input: SendableRecordBatchStream,
+        random_state: RandomState,
+        hash_key_indices: Vec<usize>,
+    ) -> DataFusionResult<Self> {
+        // Collect all batches from the input stream
+        let initial = (Vec::new(), 0);
```

Comment:
I am not sure, but If the right stream is large,  may it consume too much memory to construct the HashMap?

Reply:
Yes, It will. As noticed in the previous doc, this is something like HashJoin, which would have the same issue. There are a few things that can alleviate this:
- Limit the memory usage and fail the query if it exceeds the quota
- Hash partition this operator and run it part by part

But this operator will still require to buffer some data for building the hash table

Reply:
The second way looks good.